packages/gatsby/src/utils/node-manifest.ts

```diff
@@ -0,0 +1,302 @@
+ */
+interface INodeManifestOut {
+  page: INodeManifestPage
+  node: {
+    id: string
+  }
+  foundPageBy: FoundPageBy
+}
+
+type FoundPageBy =
+  | `ownerNodeId`
+  | `context.id`
+  | `queryTracking`
+  | `filesystem-route-api`
+  | `none`
+
+/**
+ * Finds a final built page by nodeId
+ *
+ * Note that this function wont work properly in `gatsby develop`
+ * since develop no longer runs all page queries when creating pages.
+ * We use the node id to query mapping to find the right page but
+ * this mapping only exists once you've visited a page in your browser.
+ * When this fn is being used for routing to previews the user wont necessarily have
+ * visited the page in the browser yet.
+ */
+async function findPageOwnedByNodeId({
+  nodeId,
+}: {
+  nodeId: string
+}): Promise<{
+  page: INodeManifestPage
+  foundPageBy: FoundPageBy
+}> {
+  const state = store.getState()
+  const { pages, nodes } = state
+  const { byNode } = state.queries
+
+  // in development queries are run on demand so we wont have an accurate nodeId->pages map until those pages are visited in the browser. We want this mapping before the page is visited in the browser so we can route to the right page in the browser.
+  // So in development we can just use the Map of all pages (pagePath -> pageNode)
+  // but for builds (preview inc builds or regular builds) we will have a full map
+  // of all nodeId's to pages they're queried on and we can use that instead since it
+  // will be a much smaller list of pages, resulting in better performance for large sites
+  const usingPagesMap: boolean =
+    !!process.env._GATSBY_INTERNAL_TEST_NODE_MANIFEST_AS_DEVELOP ||
+    `development` === process.env.NODE_ENV
+
+  const pagePathSetOrMap = usingPagesMap
+    ? // this is a Map of page path to page node
+      pages
+    : // this is a Set of page paths
+      byNode?.get(nodeId)
+
+  // the default page path is the first page found in
+  // node id to page query tracking
+  let pagePath = byNode?.get(nodeId)?.values()?.next()?.value
+
+  let foundPageBy: FoundPageBy = pagePath ? `queryTracking` : `none`
+
+  // but if we have more than one page where this node shows up
+  // we need to try to be more specific
+  if (pagePathSetOrMap && pagePathSetOrMap.size > 1) {
+    let ownerPagePath: string | undefined
+    let foundOwnerNodeId = false
+
+    // for each page this nodeId is queried in
+    pagePathSetOrMap.forEach((...args) => {
+      if (
+        // if we haven't found a page with this nodeId
+        // set as page.ownerNodeId then run this logic.
+        // this condition is on foundOwnerNodeId instead of ownerPagePath
+        // in case we find a page with the nodeId in page.context.id
+        // and then later in the loop there's a page with this nodeId
+        // set on page.ownerNodeId.
+        // We always want to prefer ownerPagePath over context.id
+        !foundOwnerNodeId
+      ) {
```

Comment:
code-style + nitpick.

I would switch the if statement so less code is indented by an if. This is more a personal preference that makes it easier to understand the code. There is no other code path so it makes sense to show the code that bails out of it right away + it stops the loop.

Let me know if I missed something and this isn't possible

```
if (foundOwnerNodeId) {
  break;
}

// if we haven't found a page with this nodeId
// set as page.ownerNodeId then run this logic.
// this condition is on foundOwnerNodeId instead of ownerPagePath
// in case we find a page with the nodeId in page.context.id
// and then later in the loop there's a page with this nodeId
// set on page.ownerNodeId.
// We always want to prefer ownerPagePath over context.id
```